src/v/config/property.h

```diff
@@ -953,5 +963,118 @@ class hidden_when_default_property : public property<T> {
+        using type = T::value_type;
+    };
+    template<reflection::is_std_optional T>
+    struct value_type<T, std::void_t<typename T::value_type>> {
+        using type = T::value_type;
+    };
+
+    using T = typename P::underlying;
+    using ValT = typename value_type<T>::type;
+    using val_container_t = std::vector<ValT>;
+    using restrict_check_t = std::function<bool(const ValT&)>;
+
+public:
+    template<typename R, typename... Args>
+    requires std::ranges::range<R>
+               && std::is_same_v<std::ranges::range_value_t<R>, ValT>
+    enterprise(config_store& conf, R restrict, Args&&... args)
+      : P(conf, std::forward<Args>(args)...)
+      , _conf(conf)
+      , _restricted_vals(
+          std::make_move_iterator(restrict.begin()),
+          std::make_move_iterator(restrict.end())) {}
+
+    template<typename... Args>
+    enterprise(config_store& conf, ValT restrict, Args&&... args)
+      : P(conf, std::forward<Args>(args)...)
+      , _conf(conf)
+      , _restricted_vals({restrict}) {}
+
+    template<typename... Args>
+    enterprise(config_store& conf, restrict_check_t restrict, Args&&... args)
+      : P(conf, std::forward<Args>(args)...)
+      , _conf(conf)
+      , _restricted_check(restrict) {}
+
+    bool check_restricted(const T& setting) const {
+        auto chkr = _restricted_check.value_or(
+          [](const ValT&) { return false; });
+
+        auto restricted = [this](const ValT& v) -> bool {
+            return std::find(
+                     _restricted_vals.begin(), _restricted_vals.end(), v)
+                   != _restricted_vals.end();
+        };
+
+        if constexpr (detail::Array<T>) {
+            return std::any_of(
+              setting.begin(),
+              setting.end(),
+              [&restricted, &chkr](const auto& v) {
+                  return restricted(v) || chkr(v);
+              });
+        }
+        if constexpr (reflection::is_std_optional<T>) {
+            return setting.has_value()
+                   && (restricted(setting.value()) || chkr(setting.value()));
+        }
+
+        if constexpr (std::is_same_v<T, ValT>) {
+            return restricted(setting) || chkr(setting);
+        }
+    }
+
+    std::optional<validation_error> check_restricted(YAML::Node n) const final {
+        auto v = std::move(n.as<T>());
+        if (check_restricted(v)) {
+            return std::make_optional<validation_error>(
+              P::name().data(),
+              ssx::sformat(
+                "Restricted values: {{{}}} are restricted",
+                "TODO(oren): ensure formattable"
+                // fmt::join(_restricted_vals, ",")
+                ));
+        }
+        return std::nullopt;
+    }
+
+    std::string_view type_name() const {
+        return detail::property_type_name<T>();
+    }
+
+private:
+    config_store& _conf;
+    val_container_t _restricted_vals;
+    std::optional<restrict_check_t> _restricted_check;
```

Comment:
nit: from the constructors, it seems we are using one or the other.
Could we get away with using a `std::variant` here? It should also remove the need for the `optional`

Reply:
Yeah great point. I'll see what the compiler has to say about it

Reply:
done